Proper support for java-platform dependencies

**Environment**:

- *Jib version:* 2.0.0
- *Build tool:* Gradle 6.1.1, Skaffold 1.3.1
- *OS:* MacOS Catalina 10.15.3

**Description of the issue**:
In a multi-module project, I have a project using the java-platform plugin that manages a group of dependencies and then I have several other applications that depend on that project in order to get their dependencies.  When I attempt to build these docker images using skaffold I get the error in the log output below.

I attempted adding the runtimeClasspath configuration to my platform project, but then it complained about the java convention being missing.  This problem seems to stem from the jib plugin expecting the java plugin to be applied, but java-platform does not allow for that.

I have attached a sample project that reproduces this error:  [jib-issue.tar.gz](https://github.com/GoogleContainerTools/jib/files/4171390/jib-issue.tar.gz)

**Expected behavior**:

Jib should correctly calculate the dependencies of a java-platform project.

**Steps to reproduce**:
<!-- Please provide a minimal and precise series of steps -->

  1. Create a multimodule project with a java-platform project
  2. Create a skaffold file to build it
  3. Run `skaffold dev`

**`jib-gradle-plugin` Configuration**: <!-- Delete this section if not used -->
```groovy
plugins {
  id 'com.google.cloud.tools.jib' version '2.0.0' apply false
}

project('core') {
  apply plugin: 'java-library'
}

project('platform') {
  apply plugin: 'java-platform'

  javaPlatform {
    allowDependencies()
  }
  dependencies {
    api project(':core')
  }
}

project('app') {
  apply plugin: 'java'
  apply plugin: 'application'
  apply plugin: 'com.google.cloud.tools.jib'
  dependencies {
    implementation project(':platform')
  }

  application {
    mainClassName = 'jib.issue.App'
  }

  test {
    useJUnitPlatform()
  }
}
```

**Log output**: <!-- If applicable, provide relevant log output -->

```
Listing files to watch...
 - jib-issue/app
FATA[0000] watching files for artifact jib-issue/app: listing files: listing files: getting jib-gradle dependencies: initial Jib dependency refresh failed: failed to get Jib dependencies: Running [/Users/****/jib-issue/gradlew _skaffoldFailIfJibOutOfDate -Djib.requiredVersion=1.4.0 :app:_jibSkaffoldFilesV2 -q]
 - stdout:
 - stderr:
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:_jibSkaffoldFilesV2'.
> Configuration with name 'runtimeClasspath' not found.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 457ms
: exit status 1
```

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/skaffold/FilesTaskV2.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/skaffold/FilesTaskV2.java
@@ -146,19 +146,21 @@ public class FilesTaskV2 extends DefaultTask {
 
     // Add sources + resources
     JavaPluginConvention javaConvention =
-        project.getConvention().getPlugin(JavaPluginConvention.class);
-    SourceSet mainSourceSet =
-        javaConvention.getSourceSets().findByName(SourceSet.MAIN_SOURCE_SET_NAME);
-    if (mainSourceSet != null) {
-      mainSourceSet
-          .getAllSource()
-          .getSourceDirectories()
-          .forEach(
-              sourceDirectory -> {
-                if (sourceDirectory.exists()) {
-                  skaffoldFilesOutput.addInput(sourceDirectory.toPath());
-                }
-              });
+        project.getConvention().findPlugin(JavaPluginConvention.class);
+    if (javaConvention != null) {
+      SourceSet mainSourceSet =
+          javaConvention.getSourceSets().findByName(SourceSet.MAIN_SOURCE_SET_NAME);
+      if (mainSourceSet != null) {
+        mainSourceSet
+            .getAllSource()
+            .getSourceDirectories()
+            .forEach(
+                sourceDirectory -> {
+                  if (sourceDirectory.exists()) {
+                    skaffoldFilesOutput.addInput(sourceDirectory.toPath());
+                  }
+                });
+      }
     }
   }
 
@@ -177,18 +179,20 @@ public class FilesTaskV2 extends DefaultTask {
       Project currentProject = projects.pop();
 
       // Search through all dependencies
-      for (Configuration configuration :
+      Configuration runtimeClasspath =
           currentProject
               .getConfigurations()
-              .getByName(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME)
-              .getHierarchy()) {
-        for (Dependency dependency : configuration.getDependencies()) {
-          if (dependency instanceof ProjectDependency) {
-            // If this is a project dependency, save it
-            ProjectDependency projectDependency = (ProjectDependency) dependency;
-            if (!projectDependencies.contains(projectDependency)) {
-              projects.push(projectDependency.getDependencyProject());
-              projectDependencies.add(projectDependency);
+              .findByName(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME);
+      if (runtimeClasspath != null) {
+        for (Configuration configuration : runtimeClasspath.getHierarchy()) {
+          for (Dependency dependency : configuration.getDependencies()) {
+            if (dependency instanceof ProjectDependency) {
+              // If this is a project dependency, save it
+              ProjectDependency projectDependency = (ProjectDependency) dependency;
+              if (!projectDependencies.contains(projectDependency)) {
+                projects.push(projectDependency.getDependencyProject());
+                projectDependencies.add(projectDependency);
+              }
             }
           }
         }

--- a/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/TestProject.java
+++ b/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/TestProject.java
@@ -55,6 +55,7 @@ public class TestProject extends TemporaryFolder implements Closeable {
   }
 
   private final String testProjectName;
+  private String gradleVersion = JibPlugin.GRADLE_MIN_VERSION.getVersion();
   private GradleRunner gradleRunner;
 
   private Path projectRoot;
@@ -78,9 +79,14 @@ public class TestProject extends TemporaryFolder implements Closeable {
 
     gradleRunner =
         GradleRunner.create()
-            .withGradleVersion(JibPlugin.GRADLE_MIN_VERSION.getVersion())
+            .withGradleVersion(gradleVersion)
             .withProjectDir(projectRoot.toFile())
             .withPluginClasspath();
+  }
+
+  public TestProject withGradleVersion(String version) {
+    gradleVersion = version;
+    return this;
   }
 
   public BuildResult build(String... gradleArguments) {

--- a/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/skaffold/FilesTaskV2Test.java
+++ b/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/skaffold/FilesTaskV2Test.java
@@ -40,6 +40,10 @@ public class FilesTaskV2Test {
 
   @ClassRule public static final TestProject multiTestProject = new TestProject("multi-service");
 
+  @ClassRule
+  public static final TestProject platformProject =
+      new TestProject("platform").withGradleVersion("5.2");
+
   /**
    * Verifies that the files task succeeded and returns the list of paths it prints out.
    *
@@ -138,6 +142,25 @@ public class FilesTaskV2Test {
             complexServiceRoot.resolve(
                 "local-m2-repo/com/google/cloud/tools/tiny-test-lib/0.0.1-SNAPSHOT/tiny-test-lib-0.0.1-SNAPSHOT.jar")),
         result.getInputs());
+    Assert.assertEquals(result.getIgnore().size(), 0);
+  }
+
+  @Test
+  public void testFilesTask_platformProject() throws IOException {
+    Path projectRoot = platformProject.getProjectRoot();
+    Path platformRoot = projectRoot.resolve("platform");
+    Path serviceRoot = projectRoot.resolve("service");
+    SkaffoldFilesOutput result =
+        new SkaffoldFilesOutput(verifyTaskSuccess(platformProject, "service"));
+    assertPathListsAreEqual(
+        ImmutableList.of(
+            projectRoot.resolve("build.gradle"),
+            projectRoot.resolve("settings.gradle"),
+            serviceRoot.resolve("build.gradle"),
+            platformRoot.resolve("build.gradle")),
+        result.getBuild());
+    assertPathListsAreEqual(
+        ImmutableList.of(serviceRoot.resolve("src/main/java")), result.getInputs());
     Assert.assertEquals(result.getIgnore().size(), 0);
   }
 }

--- a/jib-gradle-plugin/src/test/resources/gradle/projects/platform/build.gradle
+++ b/jib-gradle-plugin/src/test/resources/gradle/projects/platform/build.gradle
@@ -1,0 +1,5 @@
+subprojects {
+  repositories {
+    mavenCentral()
+  }
+}

--- a/jib-gradle-plugin/src/test/resources/gradle/projects/platform/platform/build.gradle
+++ b/jib-gradle-plugin/src/test/resources/gradle/projects/platform/platform/build.gradle
@@ -1,0 +1,9 @@
+plugins {
+  id 'java-platform'
+}
+
+dependencies {
+  constraints {
+    api 'org.apache.commons:commons-io:1.3.2'
+  }
+}

--- a/jib-gradle-plugin/src/test/resources/gradle/projects/platform/service/build.gradle
+++ b/jib-gradle-plugin/src/test/resources/gradle/projects/platform/service/build.gradle
@@ -1,0 +1,19 @@
+plugins {
+  id 'java'
+  id 'com.google.cloud.tools.jib'
+}
+
+sourceCompatibility = 1.8
+targetCompatibility = 1.8
+
+dependencies {
+  implementation platform(project(':platform'))
+
+  implementation 'org.apache.commons:commons-io'
+}
+
+jib {
+  to {
+    image = System.getProperty("_TARGET_IMAGE")
+  }
+}

--- a/jib-gradle-plugin/src/test/resources/gradle/projects/platform/service/src/main/java/com/test/HelloWorld.java
+++ b/jib-gradle-plugin/src/test/resources/gradle/projects/platform/service/src/main/java/com/test/HelloWorld.java
@@ -1,0 +1,23 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.test;
+
+public class HelloWorld {
+  public static void main(String[] args) {
+    System.out.println("Hello world");
+  }
+}

--- a/jib-gradle-plugin/src/test/resources/gradle/projects/platform/settings.gradle
+++ b/jib-gradle-plugin/src/test/resources/gradle/projects/platform/settings.gradle
@@ -1,0 +1,2 @@
+include ':platform'
+include ':service'

